### PR TITLE
Batch updates and better API

### LINKED ISSUE #3: FR: Return debugging information from failures
When access fails I'd love to get more than `false`.  Would be good to get the full API response so that I can find unexpected errors or make sure that the error is of the kind I expected.

### LINKED ISSUE #8: Actually expose setData() and setRules() on the database type
As far as I am aware, the functions in the `README` do not exist.

Instead I had to hack up my own version:

```typescript
function clearMockData(database: any) {
  // TODO: should not have to call both .data and .collections
  database.data = undefined;
  database.collections = undefined;
  database.rules = undefined;
}

function setMockDataAndRules(database: any, data: any, rulesFile: string) {
  database.data = data;
  database.collections = data;
  database.rules = getRulesFile(rulesFile);
}
```

### LINKED ISSUE #9: API to test batched writes
This module is missing an API to test batched writes.

```js
await database.canCommit([
   firebase.Batch.set('users/userA', { .. }),
   firebase.Batch.delete('users/userA')
])
```

The goal is to have an API that match the firestore naming as much as possible:

- `batch.commit` -> `database.canCommit(batch: Array<firestore.Batch.Operation>)`
- `batch.set` -> `firestore.Batch.set(path: string, value: any)`
- `batch.delete ` -> `firestore.Batch.delete(path: string)`
- `batch.update ` -> `firestore.Batch.update(path: string, data: Object)`

The main blocker is that the Firestore rules API doesn't seem to support batch updates. Can you confirm @samtstern ?

### PR DESCRIPTION
**Pull request by SamyPesse**
- [x] Fix #9: add method to test batch updates
- [x] Fix #8: expose `setData` and `setRules`
- [x] Fix #3: Return more than a boolean and method `firestore.assert` to throw clean error

  ```js
  const result = await database.canSet({}, 'users/userA', { name: 'John' });
  firestore.assert(result)
  ```

### CODE REVIEW

File: src/database.js [hunk 1]
```diff
@@ -205,126 +234,215 @@ class Database {
...
+     * Create the mocks for the API to represent the dataset.
+     */
+    createMockFunctions(): FirestoreMockFunction[] {
+        const documents = this.getDocuments();
+        const defaults = [
+            {
+                function: 'get',
+                args: [{ anyValue: {} }],
+                result: {
+                    value: null
```
Comment (SamyPesse) [inline src/database.js]: @samtstern Can you confirm that when a document doesn't exist, `get` and `getAfter` return `null` ?

File: src/__tests__/fixtures/firestore.rules [hunk 2]
```diff
@@ -1,12 +1,30 @@
...
+          allow read: if isPublicUser(userID);
+          allow update: if isCurrentUser(userID);
+      }
+    }
+
+    match /settings/{userID} {
+        allow read: if isCurrentUser(userID);
+        allow write: if (
+            isCurrentUser(userID) &&
+            getAfter(/databases/$(database)/documents/users/$(userID)) != null
```
Comment (SamyPesse) [inline src/__tests__/fixtures/firestore.rules]: Related to my question about result of `get` and `getAfter`, is that the right way that the other document exist ?

Is there a `existAfter` function ?

File: README.md [hunk 3]
```diff
@@ -48,32 +46,50 @@ await database.authorize();
...
-```
+Run `get`, `set`, `update` and `commit` tests using:
+
+- `database.canGet(auth: FirestoreAuth, document: string): Promise<TestResult>`
+- `database.cannotGet(auth: FirestoreAuth, document: string): Promise<TestResult>`
+- `database.canSet(auth: FirestoreAuth, document: string, value: any): Promise<TestResult>`
+- `database.cannotSet(auth: FirestoreAuth, document: string, value: any): Promise<TestResult>`
+- `database.canUpdate(auth: FirestoreAuth, document: string, values: Object): Promise<TestResult>`
+- `database.cannotUpdate(auth: FirestoreAuth, document: string, values: Object): Promise<TestResult>`
+- `database.canCommit(auth: FirestoreAuth, batch: BatchOperation[]): Promise<TestResult>`
```
Comment (samtstern) [inline README.md]: Is `cannotCommit` missing here?

File: README.md [hunk 4]
```diff
@@ -48,32 +46,50 @@ await database.authorize();
...
+- `database.cannotUpdate(auth: FirestoreAuth, batch: firestore.BatchOperation[]): Promise<TestResult>`
+
+Control test testing environment using:
+
+- `database.setData(data: FirestoreCollections)`: Update the dataset
+- `database.setRules(rules: string)`: Update the rules being tested
+- `database.setRulesFromFile(file: string)`: Read the rules from a file
```
Comment (samtstern) [inline README.md]: Nice!

File: src/batch.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+function update(document: string, data: { [string]: any }): OperationUpdate {
+    return {
+        method: 'update',
+        document,
+        data
+    };
+}
+
+function del(document: string): OperationDelete {
```
Comment (samtstern) [inline src/batch.js]: nit: might as well name this `delete` to be consistent with the underlying firestore stuff.

File: src/database.js (hunk 1 above)
Comment (samtstern) [reply src/database.js]: Yes that's correct, both return a `null` resource object.

File: src/__tests__/fixtures/firestore.rules (hunk 2 above)
Comment (samtstern) [reply src/__tests__/fixtures/firestore.rules]: No but `existAfter` would be nice!  I'll ask about that,

File: src/batch.js (hunk 5 above)
Comment (samtstern) [reply src/batch.js]: :+1:

**Comment by samtstern** [comment]
@SamyPesse this looks great, let me know when it's released and I'll bring it in to my project.

**Pull request merged**